### PR TITLE
fix(deepseek): default DeepSeek V4 to HybridEP

### PR DESCRIPTION
**Pull request by cuichenx**
# What does this PR do?

Default Bridge-created DeepSeek-V4 providers to the `flex`/HybridEP MoE dispatcher with 16 dispatcher SMs and unfused HybridEP permutation. This avoids the standard unequal-split all-to-all path that produced non-finite logits and degenerate token-0 generation in the affected multi-node Hopper inference configuration.

Fixes NVBug 6595026.

Related to #5462.

# Changelog

- Set the DeepSeek-V4 provider's dispatcher defaults to `flex`/HybridEP, 16 dispatcher SMs, and unfused HybridEP permutation.
- Move DeepSeek-V4 GB200/GB300 recipe SM overrides to the current unified field so their independently tuned 16/32-SM values retain precedence.
- Document HybridEP runtime and topology auto-detection/override behavior in the DeepSeek-V4 example.
- Add focused assertions for the provider defaults and final hardware/performance recipe configurations.

# Scope

The provider default applies to newly constructed Bridge DeepSeek-V4 providers, including the direct-HF inference path. Native Megatron checkpoints retain their serialized dispatcher configuration. Library recipes also retain their explicit, independently qualified overrides: H100 and GB300 library recipes remain on `alltoall`, while GB200 library recipes remain on HybridEP. Existing GB200/GB300 HybridEP performance recipes retain their tuned 32-SM setting.

The generic text-generation utility is unchanged. The repository still contains intentional `alltoall` configurations because token dropping and MoE Multi-LoRA require the standard dispatcher, and because dispatcher support and performance depend on the installed runtime, hardware, and topology.

# Validation

- Full DeepSeek-V4-Flash-Base inference, 2 nodes / 16 H100 GPUs, TP1/PP1/EP16/ETP1, 20 generated tokens, using the provider default without dispatcher CLI overrides and with `NCCL_NET` and all three optional HybridEP topology overrides unset:
  - runtime resolved `flex`/HybridEP, 16 dispatcher SMs, and unfused HybridEP permutation;
  - first-five variances were `25.2412, 26.2589, 27.4498, 28.1335, 28.3474`, token IDs were `1730, 696, 851, 1346, 538`, and the completion was coherent and matched the known-good control.
- Focused DeepSeek-V4 provider, library-recipe, and performance-recipe tests: 86 passed.
- `uv run --no-sync pre-commit run --all-files`: passed locally and in the exact candidate container.
- `git diff --check`, Python compile checks, and `bash -n examples/models/deepseek_v4/inference.sh`: passed.

The controlled end-to-end validation covers full-model inference only; this PR does not make a training convergence or performance claim.

# Before your PR is "Ready for review"

- [x] Read and followed the contributor guidelines.
- [x] Added focused regression tests.
- [x] Updated the DeepSeek-V4 example documentation.
- [x] The change selects the container-provided HybridEP runtime but adds no direct optional-package import or dependency.

This change was developed with assistance from OpenAI Codex.

### CODE REVIEW

File: src/megatron/bridge/perf_recipes/deepseek/gb200/deepseek_v4.py [hunk 1]
```diff
@@ -60,6 +59,8 @@ def deepseek_v4_flash_pretrain_128gpu_gb200_fp8mx_config() -> ConfigContainer:
     cfg.model.fine_grained_offloading_max_inflight_offloads = None
 
     _benchmark_common(cfg, cross_entropy_impl="native")
+    cfg.model.moe_flex_dispatcher_num_sms = 32
+    cfg.model.moe_hybridep_num_sms = None
```
Comment (malay-nagda) [inline src/megatron/bridge/perf_recipes/deepseek/gb200/deepseek_v4.py:63]: @cuichenx can we apply this change also for GB300 DSv4 flash?

cc: @dingqingy-nv

File: src/megatron/bridge/perf_recipes/deepseek/gb300/deepseek_v4.py [hunk 2]
```diff
@@ -72,6 +72,8 @@ def deepseek_v4_pro_pretrain_256gpu_gb300_fp8mx_config() -> ConfigContainer:
     cfg.model.recompute_modules = ["mla_up_proj", "mhc"]
 
     _benchmark_common(cfg, cross_entropy_impl="native")
+    cfg.model.moe_flex_dispatcher_num_sms = 32
+    cfg.model.moe_hybridep_num_sms = None
```
Comment (malay-nagda) [inline src/megatron/bridge/perf_recipes/deepseek/gb300/deepseek_v4.py:76]: @cuichenx can we apply this change also for GB200 DSv4 pro?

cc: @dingqingy-nv

File: examples/conversion/hf_to_megatron_generate_text.py [hunk 3]
```diff
@@ -161,6 +161,29 @@ def _decode_completion(tokenizer, generated_ids: torch.Tensor, prompt_length: in
...
+        "moe_token_dispatcher_type": "flex",
+        "moe_flex_dispatcher_backend": args.moe_flex_dispatcher_backend,
+    }
+    if args.moe_flex_dispatcher_num_sms is not None:
+        overrides["moe_flex_dispatcher_num_sms"] = args.moe_flex_dispatcher_num_sms
+    if args.hybridep_permute_fusion is not None:
+        if args.moe_flex_dispatcher_backend != "hybridep":
+            raise ValueError("--[no-]hybridep-permute-fusion requires --moe-flex-dispatcher-backend=hybridep")
+        overrides["moe_permute_fusion_into_hybridep"] = args.hybridep_permute_fusion
+    return overrides
```
[the remark this answers is not in the document]
Comment (cuichenx) [reply examples/conversion/hf_to_megatron_generate_text.py]: Thanks — this is stale after the scope cleanup. The PR no longer changes the generic generator or constructs a flex-dispatcher override there; DeepSeek-V4 now receives the dispatcher choice from its provider. Current MCore explicitly permits `moe_shared_expert_overlap` with the `flex` dispatcher, and the resulting provider configuration was exercised end to end with finite, coherent output. No generator override is needed.

File: src/megatron/bridge/perf_recipes/deepseek/gb200/deepseek_v4.py (hunk 1 above)
Comment (cuichenx) [reply src/megatron/bridge/perf_recipes/deepseek/gb200/deepseek_v4.py:63]: This already applies to GB300 DeepSeek-V4 Flash through inheritance: `deepseek_v4_flash_pretrain_128gpu_gb300_fp8mx_config()` starts from the GB200 Flash performance config, so it retains `moe_flex_dispatcher_num_sms = 32` and the cleared legacy field. `test_deepseek_v4_flash_128gpu_gb300_fp8mx_config` explicitly asserts both effective values.

File: src/megatron/bridge/perf_recipes/deepseek/gb300/deepseek_v4.py (hunk 2 above)
Comment (cuichenx) [reply src/megatron/bridge/perf_recipes/deepseek/gb300/deepseek_v4.py:76]: There is currently no GB200 DeepSeek-V4 Pro recipe in this PR or on `main`; the only DeepSeek-V4 Pro recipe is the GB300 one. Adding a new GB200 performance recipe here would require a separately qualified GB200 configuration and runtime evidence, so I have not guessed those settings or expanded this fix. I am leaving this thread open pending clarification or a validated GB200 Pro recipe.

**Review by malay-nagda — APPROVED** [review]
